pkg/tempopb/prealloc.go

```diff
@@ -0,0 +1,45 @@
+package tempopb
+
+import (
+	"github.com/prometheus/prometheus/pkg/pool"
+)
+
+var (
+	// buckets: [1KiB, 2KiB, 4KiB, 8KiB, 16KiB]
+	bytePool = pool.New(1_000, 16_000, 2, func(size int) interface{} { return make([]byte, 0, size) })
```

Comment:
are these numbers backed up with metrics?

Reply:
To some extent yes. We don't have the exact batch level metrics on ingesters, but noting down what I had in mind -

Internally, we are seeing ~180MiB/sec and ~1250 batches/sec. So roughly 144KiB per batch.
50th percentile of `tempo_distributor_traces_per_batch_bucket` is 20.
That's roughly 7.2KiB per trace per batch.

Reply:
Ha, this is hard.  Ok I'm getting:
- 156 MB/s to the distribs * 3 RF = 470 MB/s of proto to the ingesters
   `(sum(rate(tempo_distributor_bytes_received_total{cluster="ops-us-east-0"}[10m])) / 1024 / 1024) * 3`
- 140k ingester pushes/second
   `sum(rate(tempo_distributor_ingester_appends_total{cluster="ops-us-east-0"}[30m]))`
   `sum(rate(tempo_request_duration_seconds_count{cluster="ops-us-east-0", route="/tempopb.Pusher/PushBytes"}[30m]))`

470 MB/s / 140k pushes/s = 3.3 MB/push

- p50 of traces/batch = 20
   `histogram_quantile(.5, sum(rate(tempo_distributor_traces_per_batch_bucket{cluster="ops-us-east-0"}[30m])) by (le))`

3.3 MB / 20 = 168KB?  

Close enough for me!

Reply:
Whoa, if your calculation is right then I'll have to change my buckets because I don't have a bucket that can store 168KB traces and we'd basically be creating a new one each time = no improvement.

I agree with the first half of your calculation, we should be calculating ingester level bytes received. However, in the second half, I think there are two changes needed:


1. > 470 MB/s / 140k pushes/s = 3.3 MB/push

this would be  470MB/s / 140000 pushes/s ~ 3.3KB/push


2.  > p50 of traces/batch = 20
histogram_quantile(.5, sum(rate(tempo_distributor_traces_per_batch_bucket{cluster="ops-us-east-0"}[30m])) by (le))

This metric is for incoming batches at the distributor right? Not at the ingester. 

```

                                   (smaller batches)  -- ingester-0
                                 /
agent -- (batches) --> distributor 
                                 \
                                   (smaller batches)  -- ingester-1

```

20 traces/batch at distributor = 20 * 3 (RF) / 40 (# of ingesters) = 1.5 traces/batch at ingester

3.3KB/1.5 ~2KB per trace per batch at the ingester. which sounds closer to our avg trace size (3KB)

I think I should adjust my buckets to [0.25KB, 0.5KiB, 1KiB, 2KiB, 4KiB, 8KiB]. WDYT?